pkg/network/http/ebpf_gotls.go

```diff
@@ -158,77 +252,122 @@ func supportedArch(arch string) bool {
 	return arch == string(bininspect.GoArchX86_64)
 }
 
-func (p *GoTLSProgram) handleNewBinary(binPath string) {
-	f, err := os.Open(binPath)
+func (p *GoTLSProgram) handleProcessStart(pid pid) {
+	exePath := filepath.Join(p.procRoot, strconv.FormatUint(uint64(pid), 10), "exe")
+	binPath, err := os.Readlink(exePath)
 	if err != nil {
-		log.Errorf("Could not open file %q due to %s", binPath, err)
+		log.Debugf(" could not read binary path for pid %d: %s", pid, err)
 		return
 	}
-	defer f.Close()
-	elfFile, err := elf.NewFile(f)
-	if err != nil {
-		log.Errorf("File %q could not be parsed as elf: %s", binPath, err)
+
+	var stat syscall.Stat_t
+	if err = syscall.Stat(binPath, &stat); err != nil {
+		log.Errorf("could not stat binary path %s: %s", binPath, err)
 		return
 	}
 
-	result, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
-	if err != nil {
-		log.Errorf("Failed inspecting binary %q: %s", binPath, err)
+	hookedBin := p.analysisCache.Get(stat.Ino)
+	if hookedBin == nil {
+		f, err := os.Open(binPath)
+		if err != nil {
+			log.Errorf("could not open file %s, %s", binPath, err)
+			return
+		}
+		defer f.Close()
+
+		elfFile, err := elf.NewFile(f)
+		if err != nil {
+			log.Errorf("file %s could not be parsed as an ELF file: %s", binPath, err)
+			return
+		}
+
+		inspectionResult, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
+		if err != nil {
+			if !errors.Is(err, binversion.ErrNotGoExe) {
+				log.Errorf("error reading exe: %s", err)
+			}
+			return
+		}
+
+		if err = p.addInspectionResultToMap(inspectionResult, stat.Ino); err != nil {
+			log.Error(err)
+			return
+		}
+
+		probeIDs, err := p.attachHooks(inspectionResult, binPath)
+		if err != nil {
+			log.Errorf("error while attaching hooks: %s", err)
+			p.removeInspectionResultFromMap(stat.Ino)
+			return
+		}
+		log.Debugf("attached hooks on %s (%d)", binPath, stat.Ino)
+		hookedBin = p.analysisCache.Set(stat.Ino, &hookedBinary{probeIDs, make(runningProcessesSet)}, ttlcache.NoTTL)
+	} else {
+		log.Debugf("resetting TTL on %s (%d)", binPath, stat.Ino)
+		hookedBin = p.analysisCache.Set(stat.Ino, hookedBin.Value(), ttlcache.NoTTL)
+	}
+
+	hookedBin.Value().running_processes[pid] = struct{}{}
+	p.pidToIno[pid] = stat.Ino
+}
+
+func (p *GoTLSProgram) handleProcessStop(pid pid) {
+	ino, ok := p.pidToIno[pid]
+	if !ok {
 		return
 	}
+	delete(p.pidToIno, pid)
 
-	// result and bin path are being passed as parameters as a preparation for the future when we will have a process
-	// watcher, so we will run on more than one binary in one goTLSProgram.
-	if err := p.addInspectionResultToMap(result, binPath); err != nil {
-		log.Errorf("error in adding inspection result to map: %s", err)
+	hookedBin := p.analysisCache.Get(ino)
+	if hookedBin == nil {
+		log.Error("could not retrieve binary entry from cache")
```

Comment:
why is that an error?
you can easily get here is a process is terminated after 2 minutes

Reply:
> you can easily get here is a process is terminated after 2 minutes
No, because the cache entry will not have a TTL while there are still running processes associated to that entry. The TTL is set only when there are no processes with that binary. The `pidToIno` map contains the PID of running processes, so having an entry in it for the given PID but nothing in the cache is a bug, and an error we want to report.

Reply:
every entry should have a TTL
otherwise, our memory consumption will grow indefinably...
let's talk by heart about that